### DISCUSSION TITLE
Increasing memory consumption on requests exceeding max payload size

### DISCUSSION THREAD

**Discussion by pmajocha** [actix/actix-web#3468] [Q&A]
Hi,

I have a limit of 6 Mi request body size. I've noticed, that for requests coming below that limit, the memory level is steady. When I bombard the endpoint with requests of 12 Mi, the memory footprint starts increasing. It starts increasing quite rapidly, then slows down, maybe it settles on some point (after some time memory consumption stopped at 1.2Gi, while usually I have 150 Mi usage on this particular service). I tried to reproduce it locally and it behaved the same. With requests below limit it stabilized pretty quickly, but with requests above the limit it keeps increasing. The example:

```rust
async fn handler(
    _request: HttpRequest,
    _payload: Bytes,
) -> impl Responder {
    println!("handled");
    HttpResponse::BadRequest().body("ok")
}

pub async fn run() {
    HttpServer::new(move || App::new()
        .app_data(PayloadConfig::new(6 * 1024 * 1024))
        .service(web::resource("/handle").route(web::post().to(handler)))
    )
    .bind("0.0.0.0:9910")
    .unwrap()
    .run()
    .await
    .unwrap()
}
```

the client code
```rust
#[tokio::main]
async fn main() {
    let client = Client::new();
    let body = repeat("x").take(12 * 1024 * 1024).collect::<String>();
    let _resp = client.post("http://0.0.0.0:9910/handle")
        .body(body.clone())
        .send()
        .await;
    println!("{_resp:?}");
}
```

Is this expected behavior due to some caching? Can one change the behavior to free up the memory?

Thanks for responses in advance

**Comment by Samielakkad** [comment]
A few things can be happening here, and they are easy to mix together:

1. `Bytes` is a buffering extractor

With this handler:

```rust
async fn handler(_request: HttpRequest, _payload: Bytes) -> impl Responder
```

Actix has to read/aggregate the request body before your handler runs. `PayloadConfig::new(6 * 1024 * 1024)` limits how much the extractor should accept, but under load you can still have memory proportional to roughly:

```text
number of concurrent in-flight requests ? allowed payload size
```

So oversized requests can still cause a large temporary allocation pattern even though they eventually become payload-limit errors.

2. RSS often does not fall back immediately

Even after request buffers are dropped, the process RSS may stay high because the allocator keeps memory arenas for reuse instead of returning them to the OS immediately. That can look like a leak in simple tests. A better signal is whether memory keeps growing without bound after repeated waves, or whether it reaches a high-water mark and then gets reused.

3. Reject early when `Content-Length` is known

If you want to avoid reading any body that is obviously too large, add an early middleware/guard that checks `Content-Length` before the extractor buffers the body:

```rust
use actix_web::{
    body::EitherBody,
    dev::{Service, ServiceRequest, ServiceResponse},
    http::header,
    Error, HttpResponse,
};

const MAX_BODY: u64 = 6 * 1024 * 1024;

.wrap_fn(|req: ServiceRequest, srv| {
    let too_large = req
        .headers()
        .get(header::CONTENT_LENGTH)
        .and_then(|v| v.to_str().ok())
        .and_then(|s| s.parse::<u64>().ok())
        .is_some_and(|len| len > MAX_BODY);

    if too_large {
        let res = req.into_response(HttpResponse::PayloadTooLarge().finish());
        Box::pin(async { Ok(res.map_into_right_body()) })
    } else {
        let fut = srv.call(req);
        Box::pin(async move { fut.await.map(ServiceResponse::map_into_left_body) })
    }
})
```

That only helps when the client sends `Content-Length`. For chunked uploads or clients that lie, the extractor/stream still needs to enforce limits while reading.

4. For endpoints that do not need the full body, stream it

If you do not need the whole body in memory, avoid `Bytes` and consume `web::Payload` as a stream, enforcing your own counter and returning 413 once the total crosses the limit. That gives you more direct control over how much is buffered at a time.

5. Make sure you are measuring failed extractor responses

When the payload limit is exceeded, your handler should not print `handled`; the extractor should fail before the handler is called. If the handler is still called for 12 MiB bodies with a 6 MiB `PayloadConfig`, double-check that the config is attached to the same `App`/scope/resource that owns the route.

So: some increase up to a high-water mark can be expected due to buffering + allocator behavior, especially with concurrent oversized requests. If it grows without bound after concurrency stabilizes, that is more suspicious and would be worth reducing to a minimal repro that reports concurrency, request rate, OS allocator, and whether `Content-Length` is present.
